### PR TITLE
Use typed nodes in the IR code

### PR DESCRIPTION
**Pull request by lucteo**
IR code used in a few occasions untyped nodes. Use typed nodes instead.

This is a continuation of the effort to use typed nodes instead of pairing AST nodes with TypedProgram.

### CODE REVIEW

File: Sources/Core/TypedNode.swift [hunk 1]
```diff
@@ -135,10 +135,17 @@ extension TypedNode where ID: DeclID {
   }
 
   /// The type of the declared entity.
+  /// In most cases, the type is valid, so we extract the value from the optional by default.
```
Comment (kyouko-taiga) [inline Sources/Core/TypedNode.swift:138]: I feel like this comment should be an implementation note in the body of this accessor.

File: Sources/Core/TypedNode.swift [hunk 2]
```diff
@@ -135,10 +135,17 @@ extension TypedNode where ID: DeclID {
...
   public var type: AnyType {
     program.declTypes[id]!
```
Comment (kyouko-taiga) [inline Sources/Core/TypedNode.swift:140]: If we do as below, we don't need forced unwrapping anymore and we can represent failures to type check an expression with the error type.

```suggestion
    program.declTypes[id] ?? .error
```

**Review by kyouko-taiga — CHANGES_REQUESTED** [review]
Good work.

I think we can avoid exposing the type of a typed expression node as an optional.

File: Sources/Core/TypedNode.swift (hunk 2 above)
Comment (lucteo) [reply Sources/Core/TypedNode.swift:140]: Makes a lot of sense. Will fix.

**Review by kyouko-taiga — APPROVED** [review]
LGTM

**Pull request merged**